transport/src/main/java/io/netty/bootstrap/ChannelInitializerExtensions.java

```diff
@@ -51,66 +52,83 @@ static ChannelInitializerExtensions getExtensions() {
-            logger.log(logLevel, "Loaded extension: {}", extension.getClass());
-            extensions.add(extension);
-        }
-        if (!extensions.isEmpty()) {
-            Collections.sort(extensions, new Comparator<ChannelInitializerExtension>() {
-                @Override
-                public int compare(ChannelInitializerExtension a, ChannelInitializerExtension b) {
-                    return Double.compare(a.priority(), b.priority());
-                }
-            });
-            return Collections.unmodifiableList(extensions);
-        }
-        return Collections.emptyList();
-    }
-
-    /**
-     * @return {@code true} if there are no extensions, otherwise {@code false}.
-     */
-    abstract boolean isEmpty();
-
     /**
      * Get the list of available extensions. The list is unmodifiable.
      */
-    abstract Collection<ChannelInitializerExtension> extensions();
-
-    private static final class LoadedExtensions extends ChannelInitializerExtensions {
-        private final Collection<ChannelInitializerExtension> extensions;
+    abstract Collection<ChannelInitializerExtension> extensions(ClassLoader cl);
 
-        private LoadedExtensions(Collection<ChannelInitializerExtension> extensions) {
-            this.extensions = extensions;
+    private static final class EmptyExtensions extends ChannelInitializerExtensions {
+        @Override
+        Collection<ChannelInitializerExtension> extensions(ClassLoader cl) {
+            return Collections.emptyList();
         }
+    }
 
-        @Override
-        boolean isEmpty() {
-            return extensions.isEmpty();
+    private static final class ServiceLoadingExtensions extends ChannelInitializerExtensions {
+        private final InternalLogLevel logLevel;
+        private final boolean loadAndCache;
+
+        private WeakReference<ClassLoader> classLoader;
+        private Collection<ChannelInitializerExtension> extensions;
+
+        ServiceLoadingExtensions(InternalLogLevel logLevel, boolean loadAndCache) {
+            this.logLevel = logLevel;
+            this.loadAndCache = loadAndCache;
         }
 
         @SuppressWarnings("AssignmentOrReturnOfFieldWithMutableType")
         @Override
-        Collection<ChannelInitializerExtension> extensions() {
+        synchronized Collection<ChannelInitializerExtension> extensions(ClassLoader cl) {
+            ClassLoader context = Thread.currentThread().getContextClassLoader();
+            ClassLoader configured = classLoader == null ? null : classLoader.get();
+            if (configured == null || configured != cl && configured != context) {
+                Collection<ChannelInitializerExtension> loaded = serviceLoadExtensions(logLevel, cl);
+                if (loaded.isEmpty()) {
+                    // If we found no extensions, then try the context-specific class loader as well.
+                    // Frameworks like https://aries.apache.org/modules/spi-fly.html use the context class loader.
```

Comment:
I don't think it set the context class loader to make this work. I think it injected itself into the `ServiceLoader.load(Class)` callsites via bytecode rewriting.

We might not need the workaround, as it looks like spi-fly now works in the ClassLoader case.
https://github.com/apache/aries/commit/a97cb627323a59379e7b08166c2ec7c06027de34

I suggest we ditch caring about spi-fly and see what happens.